File: DailyTemperatures.java
import java.util.Stack;

public class DailyTemperatures {
    // TC: O(N) where N is length of temperatures
    // SC: O(N) where N is length of temperatures
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int[] result = new int[n];
        Stack<Integer> decStack = new Stack<>();
        for (int i = 0; i < n; i++) {
            while (!decStack.isEmpty() && temperatures[decStack.peek()] < temperatures[i]) {
                int index = decStack.pop();
                result[index] = i - index;
            }
            decStack.push(i);
        }
        return result;
    }
}

File: Daily Temperatures .py

# STACK approach        (storing -- unresolved elements in stack)
# time : O(2N), better than bruteforce O(N*N) 
# space : O(1) 
# run on leetcode: yes


class Solution:
    def dailyTemperatures(self, T):
        result = [0]*len(T)

        # EDGE CASE
        if T == [] or len(T) == 0:
            return result

        st = [0]
        for i in range(0, len(T)):
            while st != [] and T[i] > T[st[-1]]:
                idx = st.pop()
                result[idx] = i - idx 
            st.append(i)
        return result


        
obj = Solution()
print(obj.dailyTemperatures(T = [73, 74, 75, 71, 69, 72, 76, 73]))
